Add documentation on retry (#150)
edgedb/edgedb-python: docs/api/asyncio_con.rst

        Example:

        .. code-block:: python

            async for tx in con.retry():
                with tx:
                    value = tx.query_one("SELECT Counter.value")
                    tx.execute(
                        "UPDATE Counter SET { value := <int64>$value",
                        value=value,
                    )

        Note that we are executing queries on the ``tx`` object rather
        than on the original pool itself.

    .. py:method:: try_transaction()

        Execute a non-retryable transaction.

        Contrary to ``retry()``, ``try_transaction()`` will not attempt
        to re-run the nested code block in case a retryable error happens.

        This is a low-level API and it is advised to use the ``retry()``
        method instead.

        A call to ``try_transaction()`` returns
        :py:class:`AsyncIOTransaction`.

        Example:

        .. code-block:: python

            with con.try_transaction() as tx:
                value = tx.query_one("SELECT Counter.value")
                tx.execute(
                    "UPDATE Counter SET { value := <int64>$value",
                    value=value,
                )

>         Note executing queries on ``tx`` object rather than the original
>         connection.

Reviewer: ```suggestion
        Note that we are executing queries on the ``tx`` object,
        rather than on the original connection `con`.
```

    .. py:method:: transaction(isolation=None, readonly=None, deferrable=None)
        **Deprecated**. Use :py:meth:`retry` or :py:meth:`try_transaction`.

        Create a :py:class:`AsyncIOTransaction` object.

        :param isolation:
            Transaction isolation mode, can be one of:
            ``'serializable'``, ``'repeatable_read'``.  If not specified,
            the server-side default is used.

        :param readonly:
            Specifies whether or not this transaction is read-only.  If not
            specified, the server-side default is used.

        :param deferrable:
            Specifies whether or not this transaction is deferrable.  If not

>             specified, the server-side default is used.

Reviewer: I'd also link our transaction docs here that explain what readonly or deferrable do.

Author: This is deprecated part. Let's add this to the `with_transaction_options`.

            If the results of *query* are desired, :py:meth:`query` or
            :py:meth:`query_one` should be used instead.

    .. py:method:: retry()

>         Create a :py:class:`AsyncIORetry` object. This is a way to make
>         reliable transactions. See
>         :ref:`edgedb-python-asyncio-api-transaction` for more info.

Reviewer: I'd make this description a bit more high level:

```suggestion
        Execute a retryable transaction.
        
        This is the preferred method of running transactions as it
        can auto-retry transactions in case a retryable [LINK] error (like
        a network disconnect) happens.
        
        It is very important to always keep in mind that the code within
        the ``retry()`` block might be executed more than one time.
        
        A ``retry()`` call returns an instance of
        :py:class:`AsyncIORetry`. 
        
        See :ref:`edgedb-python-asyncio-api-transaction` for more info.
```

We need to document which transactions are retryable in the main EdgeDB docs and link them from here.

I'd move the `See :ref:.. for more info` to the very bottom of this section, below the example.

Reviewer: cc @elprans 

Elvis take a look at how I phrased it. Any suggestions?

Reviewer 2: I prefer "retriable" instead of "retryable", because we say "transaction is retried", although the Internet seems to be split between the spellings.

As for the description, here's my take:

```suggestion
        Open a retriable transaction loop.
       
        This is the preferred method of initiating and running a database
        transaction in a robust fashion.  The `retry()` transaction loop will
        attempt to re-execute the transaction lop body if a transient error
        occurs, such as a network error or a transaction serialization error.
       
        Returns an instance of :py:class:`AsyncIORetry`.
       
        See :ref:`edgedb-python-asyncio-api-transaction` for more details.
```

Reviewer: https://english.stackexchange.com/a/305274

Reviewer 2: That isn't an accepted answer, and I don't agree that "retriable" is necessarily a form of "triable", because, again, it's "retried", not "retryed"

Reviewer: Well, the "accepted" one isn't always the right one, since it's not accepted by some "expert", but rather by the one who asked the question in the first place :)  The accepted one doesn't speculate on which spelling is more correct either, so I posted a link to the comment that does.

Reviewer: @tailhook Let's merge Elvis' version, replacing retriable with retryable (as our native speakers suggested us to)

        Example:

        .. code-block:: python

            async for tx in con.retry():
                with tx:
                    value = tx.query_one("SELECT Counter.value")
                    tx.execute(
                        "UPDATE Counter SET { value := <int64>$value",
                        value=value,
                    )

        Note executing queries on ``tx`` object rather than the original
        connection.

    .. py:method:: try_transaction()

>         Create an individual :py:class:`AsyncIOTransaction` object.
> 
>         Transaction may fail so it's often better to use :py:meth:`retry`
>         instead. This method is targeted for users who want to implement \
>         more specialized transaction retry loop.

Reviewer: ```suggestion

        Execute a non-retryable transaction.
        
        Contrary to ``retry()``, ``try_transaction()`` will not attempt
        to re-run the nested code block in case a retryable error happens.
        
        This is a low-level API and it is advised to use the ``retry()`` 
        method instead.

        A call to ``try_transaction()`` returns 
        :py:class:`AsyncIOTransaction`.
```

    :return: An instance of :py:class:`AsyncIOPool`.

>     Queries on the connection pool can be executed directly, in this
>     case every invocation some connection is picked for the duration of
>     the query:

Reviewer: ```suggestion
    The connection pool has high-level APIs to access Connection[link]
    APIs directly, without manually acquiring and releasing connections
    from the pool:
    
    * Pool.retry()
    * Pool.query()
    * Pool.query_one()
    * Pool.query_json()
    * Pool.query_one_json()
```

    .. code-block:: python

        async with edgedb.create_async_pool(user='edgedb') as pool:
            await pool.query('SELECT {1, 2, 3}')

    Transactions can be executed as well:

    .. code-block:: python

        async with edgedb.create_async_pool(user='edgedb') as pool:
            async for tx in pool.retry():
                async with tx:
                    await tx.query('SELECT {1, 2, 3}')

>     If you need to lock on specific connection, use ``async with`` block:

Reviewer: ```suggestion
    To hold on a specific connection object, use the ``pool.acquire()`` API:
```

    :return: An instance of :py:class:`AsyncIOPool`.

    The connection pool has high-level APIs to access Connection[link]
    APIs directly, without manually acquiring and releasing connections
    from the pool:

>     * Pool.retry()

Reviewer: Can we hyperlink these?

Author: Done

    :return: An instance of :py:class:`AsyncIOPool`.

    The connection pool has high-level APIs to access Connection[link]
    APIs directly, without manually acquiring and releasing connections
    from the pool:

    * :py:meth:`AsyncIOPool.retry()`
    * :py:meth:`AsyncIOPool.query()`
    * :py:meth:`AsyncIOPool.query_one()`
    * :py:meth:`AsyncIOPool.query_json()`

>     * :py:meth:`AsyncIOPool.query_one_json()`

Reviewer: Also `execute()`

    .. code-block:: python

        async with edgedb.create_async_pool(user='edgedb') as pool:
            await pool.query('SELECT {1, 2, 3}')

    Transactions can be executed as well:

    .. code-block:: python

        async with edgedb.create_async_pool(user='edgedb') as pool:
            async for tx in pool.retry():
                async with tx:
                    await tx.query('SELECT {1, 2, 3}')

>     To hold on a specific connection object, use the ``pool.acquire()`` API:

Reviewer: ```suggestion
    To hold on to a specific connection object, use the ``pool.acquire()`` API:
```

        See :py:meth:`AsyncIOConnection.execute()
        <edgedb.AsyncIOConnection.execute>` for details.

    .. py:method:: retry()
        Create a :py:class:`AsyncIORetry` object. This is a way to make
        reliable transactions. See
        :ref:`edgedb-python-asyncio-api-transaction` for more info.

        Example:

        .. code-block::python

            async for tx in pool.retry():
                with tx:
                    value = tx.query_one("SELECT Counter.value")
                    tx.execute(
                        "UPDATE Counter SET { value := <int64>$value",
                        value=value,
                    )

>         Note executing queries on ``tx`` object rather than the original
>         pool.

Reviewer: ```suggestion
        Note that we are executing queries on the ``tx`` object rather
        than on the original pool itself.
```

    .. py:method:: try_transaction()
        Create a :py:class:`AsyncIOTransaction` object.

        Creates individual transaction.

        Transaction may fail so it's often better to use :py:meth:`retry`
        instead. This method is targeted for users who want to implement \
        more specialized transaction retry loop.

        Example:

        .. code-block::python

            with pool.try_transaction() as tx:
                value = tx.query_one("SELECT Counter.value")
                tx.execute(
                    "UPDATE Counter SET { value := <int64>$value",
                    value=value,
                )

        Note executing queries on ``tx`` object rather than the original

>         pool.

Reviewer: ditto

        See :py:meth:`AsyncIOConnection.execute()
        <edgedb.AsyncIOConnection.execute>` for details.

    .. py:method:: retry()

>         Create a :py:class:`AsyncIORetry` object. This is a way to make

Reviewer: Can we please copy the text from https://github.com/edgedb/edgedb-python/pull/150/files#diff-833af12d3f57fe855ba1925cc6b4409cef2094fb7aac456676721f0560e1e435R237?

        reliable transactions. See
        :ref:`edgedb-python-asyncio-api-transaction` for more info.

        Example:

        .. code-block::python

            async for tx in pool.retry():
                with tx:
                    value = tx.query_one("SELECT Counter.value")
                    tx.execute(
                        "UPDATE Counter SET { value := <int64>$value",
                        value=value,
                    )

        Note that we are executing queries on the ``tx`` object rather
        than on the original pool itself.

    .. py:method:: try_transaction()

>         Create a :py:class:`AsyncIOTransaction` object.

Reviewer: Ditto, `Connection.try_transaction` has slightly better wording now.

    .. py:attribute:: free_size

        Number of available connections in the pool.

.. _edgedb-python-asyncio-api-transaction:

Transactions
============

> The most robust way to make transaction is using a retry loop:

Reviewer: ```suggestion
The most robust way to execute transactional code is to use 
the ``retry()`` API:
```

Reviewer 2: I think it's important to still use the word loop, so "the ``retry()`` loop API" would probably be a more informative description.  It is impossible to use "retry()" without a loop, isn't it?

Author: Yes. Added "loop"

.. code-block:: python

    async for tx in pool.retry():
        async with tx:
            await tx.execute("INSERT User { name := 'Don' }")

> Note executing queries on ``tx`` object rather than original
> connection or connection pool.

Reviewer: ```suggestion
Note that we execute queries on the ``tx`` object in the above
example, rather than on the original connection pool ``pool`` 
object.
```

> This implies that:
> 
> 1. Transaction is executed atomically
> 2. If transaction is failed for any of the number of transient errors (i.e.
>    network failure or concurrent update error) transaction is retried
> 3. If any other exception occurs, transaction is rolled back and not
>    retried.

Reviewer: ```suggestion
The ``retry()`` API guarantees that:

1. Transactions are executed atomically;
2. If a transaction is failed for any of the number of transient errors (i.e.
   a network failure or a concurrent update error), the transaction would be retried;
3. If any other, non-retryable exception occurs, the transaction is rolled back,
   and the exception is propagated, immediately aborting the ``retry()`` block.
```

> One non-obvious implication of the (3) is that whole block is retried

Reviewer: ```suggestion
A non-obvious implication of retrying transactions is that the entire 
nested code block can be re-run, including any non-querying
Python code.
```

including non-database statements, so for example:

.. code-block:: python

    async for tx in pool.retry():
        async with tx:
            user = await tx.fetch_one(
                "SELECT User { email } FILTER .login = <str>$login",
                login=login,
            )
            data = await httpclient.get(
                'https://service.local/email_info',
                params=dict(email=user.email),
            )
            user = await tx.fetch_one('''
                    UPDATE User FILTER .login = <str>$login
                    SET { email_info := <json>$data}
                ''',
                login=login,
                data=data,
            )

> Here execution of the http request be retried too. The core of the issue
> is that whenever transaction is interrupted user might have changed the
> email, so we have to redo all the work done.

Reviewer: ```suggestion
In the above example, the execution of the HTTP request would be
retried too.
```

I don't understand the "The core of the issue..." sentence. How can the user change the email while the code is iterating with `async for` here?

Author: > I don't understand the "The core of the issue..." sentence. How can the user change the email while the code is iterating with async for here?

In parallel/concurrent request for example (or a background task or whatever).

I think we should mention somehow that we retry the whole transaction not because we're lazy, but because this is how transactions must be handled. Maybe we should pick a better example, where concurrent things are more expected to happen?

Reviewer: Maybe rephrase to:

"In the above example, the execution of the HTTP request would be
retried too to guarantee that the DB state is synchronized with
the remote service state."

?

Author: Well, I this case remote service looks like readonly, so nothing synchronized. But generally I wanted to explain why we are retrying the whole transaction block rather than individual queries. Here it's a good idea to cache the value retrieved from server and only check the cache on the second retry, so actual request isn't something that has to be done.

Author: Updated the first sentence and edited the text a bit: https://github.com/edgedb/edgedb-python/pull/150/commits/fc7dc2aa3e610a1f0d2fa1bf9ba88e4a4ce870b4

> Generally it's recommended not to do potentially slow or long running
> external tasks within the transaction and split it into multiple
> ones.

Reviewer: ```suggestion
Generally it's recommended to not execute any long running
code within the transaction unless absolutely necessary.

Transactions allocate expensive server resources and having
too many concurrently running long-running transactions will
negatively impact the performance of the DB server.
```

"split it into multiple ones." -- this is not always possible. But I think we can just leave that sentence out now.

See also:
* RFC1004_
* :py:meth:`AsyncIOPool.retry()`
* :py:meth:`AsyncIOPool.try_transaction()`
* :py:meth:`AsyncIOConnection.retry()`
* :py:meth:`AsyncIOConnection.try_transaction()`

.. py:class:: AsyncIOTransaction

>     Represents a transaction or savepoint block.

Reviewer: ```suggestion
    Represents a transaction or a savepoint block.
```

>     Transactions are created by calling the
>     :py:meth:`AsyncIOConnection.transaction()` method.

Reviewer: ```suggestion
    Instances of this type are created by calling the
    :py:meth:`AsyncIOConnection.try_transaction()` method.
```

    .. py:coroutinemethod:: start()

>         Enter the transaction or savepoint block.

Reviewer: ```suggestion
        Start a transaction or create a savepoint.
```

    .. py:attribute:: free_size

        Number of available connections in the pool.

.. _edgedb-python-asyncio-api-transaction:

Transactions
============

The most robust way to execute transactional code is to use
the ``retry()`` loop API:

.. code-block:: python

    async for tx in pool.retry():
        async with tx:
            await tx.execute("INSERT User { name := 'Don' }")

Note that we execute queries on the ``tx`` object in the above
example, rather than on the original connection pool ``pool`` 
object.

The ``retry()`` API guarantees that:

1. Transactions are executed atomically;
2. If a transaction is failed for any of the number of transient errors (i.e.
   a network failure or a concurrent update error), the transaction would be retried;
3. If any other, non-retryable exception occurs, the transaction is rolled back,
   and the exception is propagated, immediately aborting the ``retry()`` block.

The key implication of retrying transactions is that the entire 
nested code block can be re-run, including any non-querying
Python code.

> including non-database statements, so for example:

Reviewer: dangling part of old sentence here

Author: Fixed

    .. py:attribute:: free_size

        Number of available connections in the pool.

.. _edgedb-python-asyncio-api-transaction:

Transactions
============

> The most robust way to execute transactional code is to use
> the ``retry()`` loop API:

Reviewer: ```suggestion
The most robust way to execute transactional code is to use
the ``retry()`` API:
```

Reviewer: Actually, never mind, let's keep "loop API". I think it better communicates to the reader that hey, this weird looking construct is actually the api.

user might have the email changed (as the result of concurrent
transaction), so we have to redo all the work done.

Generally it's recommended to not execute any long running
code within the transaction unless absolutely necessary.

Transactions allocate expensive server resources and having
too many concurrently running long-running transactions will
negatively impact the performance of the DB server.

See also:

* RFC1004_
* :py:meth:`AsyncIOPool.retry()`
* :py:meth:`AsyncIOPool.try_transaction()`
* :py:meth:`AsyncIOConnection.retry()`
* :py:meth:`AsyncIOConnection.try_transaction()`

.. py:class:: AsyncIOTransaction

    Represents a transaction or a savepoint block.

    Instances of this type are created by calling the
    :py:meth:`AsyncIOConnection.try_transaction()` method.

    .. py:coroutinemethod:: start()

        Start a transaction or create a savepoint.

    .. py:coroutinemethod:: commit()

        Exit the transaction or savepoint block and commit changes.

    .. py:coroutinemethod:: rollback()

        Exit the transaction or savepoint block and discard changes.

    .. describe:: async with c:

>         start and commit/rollback the transaction or savepoint block

Reviewer: ```suggestion
        Start and commit/rollback the transaction or savepoint block
```
